File: sdk-java/dephealth-core/src/main/java/biz/kryukov/dev/dephealth/ErrorClassifier.java
package biz.kryukov.dev.dephealth;

import java.net.ConnectException;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import java.net.http.HttpTimeoutException;
import javax.net.ssl.SSLException;

/**
 * Classifies check exceptions into {@link CheckResult} with status category and detail.
 *
 * <p>Classification chain:
 * <ol>
 *   <li>{@link CheckException} with statusCategory/statusDetail</li>
 *   <li>Platform exception types (timeout, DNS, connection, TLS)</li>
 *   <li>Wrapped exception cause (recursive)</li>
 *   <li>Fallback: error/error</li>
 * </ol>
 */
public final class ErrorClassifier {

    private ErrorClassifier() {}

    /**
     * Classifies an exception into a CheckResult.
     *
     * @param err the exception to classify, or null for success
     * @return the classification result
     */
    public static CheckResult classify(Exception err) {
        if (err == null) {
            return CheckResult.OK;
        }

        // 1. CheckException with explicit classification.
        if (err instanceof CheckException ce) {
            return new CheckResult(ce.statusCategory(), ce.statusDetail());
        }

        // 2. Platform exception types.
        CheckResult platform = classifyPlatform(err);
        if (platform != null) {
            return platform;
        }

        // 3. Check wrapped cause.
        Throwable cause = err.getCause();
        if (cause instanceof Exception inner && cause != err) {
            CheckResult innerResult = classify(inner);
            if (!StatusCategory.ERROR.equals(innerResult.category())) {
                return innerResult;
            }
        }

        // 4. Fallback.
        return new CheckResult(StatusCategory.ERROR, "error");
    }

    private static CheckResult classifyPlatform(Throwable err) {
        if (err instanceof SocketTimeoutException) {
            return new CheckResult(StatusCategory.TIMEOUT, "timeout");
        }
        if (err instanceof java.util.concurrent.TimeoutException) {
            return new CheckResult(StatusCategory.TIMEOUT, "timeout");
        }
        if (err instanceof HttpTimeoutException) {
            return new CheckResult(StatusCategory.TIMEOUT, "timeout");
        }
        if (err instanceof UnknownHostException) {
            return new CheckResult(StatusCategory.DNS_ERROR, "dns_error");
        }
        if (err instanceof ConnectException) {
            return new CheckResult(StatusCategory.CONNECTION_ERROR, "connection_refused");
        }
        if (err instanceof java.net.NoRouteToHostException) {
            return new CheckResult(StatusCategory.CONNECTION_ERROR, "connection_refused");
        }
        if (err instanceof SSLException) {
            return new CheckResult(StatusCategory.TLS_ERROR, "tls_error");
        }
        return null;
    }
}

File: sdk-csharp/DepHealth.Core/ErrorClassifier.cs
using System.Net.Sockets;
using System.Security.Authentication;
using DepHealth.Exceptions;

namespace DepHealth;

/// <summary>
/// Classifies exceptions into status category and detail for metrics.
/// </summary>
public static class ErrorClassifier
{
    /// <summary>
    /// Classifies an exception (or null for success) into a <see cref="CheckResult"/>.
    /// </summary>
    public static CheckResult Classify(Exception? err)
    {
        if (err is null)
        {
            return CheckResult.Ok;
        }

        // 1. DepHealthException carries its own classification
        if (err is DepHealthException dhe)
        {
            return new CheckResult(dhe.ExceptionStatusCategory, dhe.ExceptionStatusDetail);
        }

        // 2. Platform timeout types
        if (err is TimeoutException or OperationCanceledException)
        {
            return new CheckResult(StatusCategory.Timeout, "timeout");
        }

        // 3. Socket errors
        if (err is SocketException se)
        {
            return ClassifySocketException(se);
        }

        // 4. DNS errors (System.Net.Sockets.SocketException with specific error code
        // is handled above; direct check for HttpRequestException wrapping DNS)
        if (IsDnsError(err))
        {
            return new CheckResult(StatusCategory.DnsError, "dns_error");
        }

        // 5. TLS/SSL errors
        if (err is AuthenticationException)
        {
            return new CheckResult(StatusCategory.TlsError, "tls_error");
        }

        // 6. Check inner exception
        if (err.InnerException is not null)
        {
            var inner = Classify(err.InnerException);
            if (inner.Category != StatusCategory.Error)
            {
                return inner;
            }
        }

        // 7. Fallback
        return new CheckResult(StatusCategory.Error, "error");
    }

    private static CheckResult ClassifySocketException(SocketException se)
    {
        return se.SocketErrorCode switch
        {
            SocketError.ConnectionRefused => new CheckResult(StatusCategory.ConnectionError, "connection_refused"),
            SocketError.HostNotFound or SocketError.NoData =>
                new CheckResult(StatusCategory.DnsError, "dns_error"),
            SocketError.TimedOut => new CheckResult(StatusCategory.Timeout, "timeout"),
            SocketError.HostUnreachable or SocketError.NetworkUnreachable =>
                new CheckResult(StatusCategory.ConnectionError, "connection_refused"),
            _ => new CheckResult(StatusCategory.Error, "error")
        };
    }

    private static bool IsDnsError(Exception err)
    {
        var msg = err.Message;
        return msg.Contains("No such host is known", StringComparison.OrdinalIgnoreCase)
            || msg.Contains("Name or service not known", StringComparison.OrdinalIgnoreCase)
            || msg.Contains("nodename nor servname provided", StringComparison.OrdinalIgnoreCase);
    }
}
